### PR TITLE
Twitter Post

### PR DESCRIPTION
**Pull request by dhagandev**
Can post to twitter, has the start of URL support. 
Checks that the length is below Twitter's cap.
Keys, Secrets, and Access Tokens are hidden.

### CODE REVIEW

**Comment by kev-chan** [comment]
a few high level comments before more detailed review:
- need to be able to run it without maven for this first lab
- please update readme with how to run on command line

File: src/main/java/LiUni/TwitterStatus.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        if (statusLength > TWITTER_CHAR_MAX) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static boolean statusContainsUrl(String text) {
```
Comment (kev-chan) [inline src/main/java/LiUni/TwitterStatus.java]: why is this function static? also is there a reason why it is always defaulted to false?

File: src/main/java/LiUni/TwitterStatus.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        try {
+            new URL(url).toURI();
+            return true;
+        }
+        catch (Exception e) {
+            return false;
+        }
+    }
+
+    private static String getUrl(String status) {
```
Comment (kev-chan) [inline src/main/java/LiUni/TwitterStatus.java]: same for this, if this is't needed please deleted

File: src/main/java/LiUni/TwitterTimeline.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.List;
+
+public class TwitterTimeline {
+    public void getTimeline() {
+        Twitter twitter = TwitterFactory.getSingleton();
+        try {
+            List<Status> statuses = twitter.getHomeTimeline();
+            System.out.println("Fetching your home timeline...");
+            for (Status status : statuses) {
+                System.out.println(status.getUser().getName() + ": " + status.getText());
```
Comment (kev-chan) [inline src/main/java/LiUni/TwitterTimeline.java]: please output the entire status item

File: src/main/java/LiUni/InputRequests.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        return userInput;
+    }
+
+    private static void promptTimeline(Scanner reader) {
+        System.out.println(getTimeline + validInput);
+        String userInput = getUserInput(reader);
+
+        if (userInput.equals("Y")) {
```
Comment (kev-chan) [inline src/main/java/LiUni/InputRequests.java]: should be a const

File: src/main/java/LiUni/KeyHandler.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            doc.getDocumentElement().normalize();
+
+            NodeList serviceList = doc.getElementsByTagName("service");
+            for (int i = 0; i < serviceList.getLength(); i++) {
+                Node service = serviceList.item(i);
+                Element element = (Element) service;
+                String serviceName = element.getAttribute("id");
+
+                //Ideally would use switch case based on what serviceName is
+                if (serviceName.equals("Twitter")) {
```
Comment (kev-chan) [inline src/main/java/LiUni/KeyHandler.java]: const for Twitter

File: src/main/java/LiUni/KeyHandler.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        }
+        catch (Exception e) {
+            System.out.println("Could not set up the keys. " + e.toString());
+            e.printStackTrace();
+        }
+    }
+
+    private static void setupTwitter(Element element) {
+        try {
+            BufferedWriter writer = new BufferedWriter(new FileWriter("twitter4j.properties"));
```
Comment (kev-chan) [inline src/main/java/LiUni/KeyHandler.java]: look into being able to guarantee the closure of the writer in the event an exception occurs within the try/catch block

File: src/main/java/LiUni/TwitterStatus.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import twitter4j.TwitterFactory;
+import twitter4j.StatusUpdate;
+import twitter4j.Status;
+
+import java.net.URL;
+
+
+public class TwitterStatus {
+
+    private static final int TWITTER_CHAR_MAX = 280;
```
Comment (kev-chan) [inline src/main/java/LiUni/TwitterStatus.java]: nice use of the static variables

**Comment by kev-chan** [comment]
please handle runtime exceptions with twitter4j better when bad keys are passed in. The exception is logged after the new prompt for next steps post failure

File: src/main/java/LiUni/TwitterStatus.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    public void postStatus(String text) {
+        if (textErrorCheck(text)) {
+            StatusUpdate newStatus = new StatusUpdate(text);
+            try {
+                Status status = twitter.updateStatus(newStatus);
+                System.out.println("Successfully updated status to [" + status.getText() + "].");
+            }
+            catch (TwitterException e) {
+                System.out.println("Exception occurred. " + e.getErrorMessage());
```
Comment (kev-chan) [inline src/main/java/LiUni/TwitterStatus.java]: you should output the stacktrace

File: src/main/java/LiUni/InputRequests.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    //True = Input is valid; False = Input has some error
+    private static boolean validInput(String input) {
+        if (input.equals(YES) || input.equals(NO)) {
+            return true;
+        }
+        return false;
+    }
+
+    private static void postingTweet(Scanner reader) {
```
Comment (kev-chan) [inline src/main/java/LiUni/InputRequests.java]: why are the methods static?

Comment (kev-chan) [reply src/main/java/LiUni/InputRequests.java]: postTweet would be a better name

Comment (dhagandev) [reply src/main/java/LiUni/InputRequests.java]: I didn't think of them as object specific.

I'll update them to not be static.

File: src/main/java/LiUni/TwitterStatus.java (hunk 8 above)
Comment (dhagandev) [reply src/main/java/LiUni/TwitterStatus.java]: Accidentally fixed this on the timeline branch.

File: src/main/java/LiUni/InputRequests.java (hunk 9 above)
Comment (dhagandev) [reply src/main/java/LiUni/InputRequests.java]: Just kidding, they need to be static due to the chain of being called from main.

File: README.md [hunk 10]
```diff
@@ -1 +1,26 @@
-#Lithium University Onboarding
+# Lithium University Onboarding
+
+## Getting Started ##
+TODO: Add set up instructions. (Install Java, IntelliJ, etc)
```
Comment (kev-chan) [inline README.md]: either fill in or remove the TODO

File: src/main/java/LiUni/InputRequests.java (hunk 9 above)
Comment (kev-chan) [reply src/main/java/LiUni/InputRequests.java]: can you refactor your code so it's not using a much of static calls? good practice is to not have static calls

Comment (dhagandev) [reply src/main/java/LiUni/InputRequests.java]: Well my original impression was that this file wouldn't really exist in this format in the future; I made it more for purposes of an easier testing interface. But I can try and figure out some way to reduce the static calls.

File: src/main/java/LiUni/TwitterStatus.java (hunk 8 above)
Comment (kev-chan) [reply src/main/java/LiUni/TwitterStatus.java]: my current understanding of the branches is 
post tweet branch --> look at post tweet
timeline branch --> look at timeline
please either combine the branches into one or have all of the post tweet information inside this branch.  At a high level, each branch should represent completely finished work that can be merged and tested.

Comment (dhagandev) [reply src/main/java/LiUni/TwitterStatus.java]: I understand; it was an accident that timeline material ended up on post and an accident that I had pushed that change on the wrong branch.

File: src/main/java/LiUni/LiUniDriver.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package LiUni;
+
+import java.util.Scanner;
+
+public class LiUniDriver {
+    public static void main(String args[]) {
```
Comment (kev-chan) [inline src/main/java/LiUni/LiUniDriver.java]: nice

File: src/main/java/LiUni/InputRequests.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
+package LiUni;
```
Comment (kev-chan) [inline src/main/java/LiUni/InputRequests.java]: everything looks good
please change the package name to all lowercase and you can merge after

**Pull request merged**